pyrelational/pipeline/generic_pipeline.py

```diff
@@ -0,0 +1,321 @@
+        self.iteration = 0
+
+        # Data structures for logging values of interest
+        self.performances = defaultdict(dict)
+        self.labelled_by = {}
+        self.log_labelled_by(data_manager.l_indices, tag="Initialisation")
+
+    def theoretical_performance(self, test_loader: Optional[DataLoader] = None) -> Dict:
+        """Returns the performance of the full labelled dataset against the
+        test data. Typically used for evaluation to establish theoretical benchmark
+        of model performance given all available training data is labelled. The
+        "horizontal" line in area under learning curve plots for active learning
+
+        Would not make much sense when we are doing active learning for the real
+        situation, hence not part of __init__
+
+        :param test_loader: Pytorch Data Loader with
+                test data compatible with model, optional as often the test loader can be
+                generated from data_manager but is here for case when it hasn't been defined
+                or there is a new test set.
+
+        :return: performances
+        """
+        self.model.train(self.train_loader, self.valid_loader)
+
+        # use test loader in data_manager if there is one
+        result = self.model.test(self.test_loader if test_loader is None else test_loader)
+        if self.data_manager.top_unlabelled is not None:
+            result["hit_ratio"] = np.nan
+        self.performances["full"] = result
+        # make sure that theoretical best model is not stored
+        self.model.current_model = None
+        return self.performances["full"]
+
+    def current_performance(self, test_loader: Optional[DataLoader] = None, query: Optional[List[int]] = None) -> Dict:
+        """
+        Current performance of model
+        :param test_loader: Pytorch Data Loader with
+                test data compatible with model, optional as often the test loader can be
+                generated from data_manager but is here for case when it hasn't been defined
+                or there is a new test set.
+
+        :param query: List of indices selected for labelling. Used for calculating hit ratio metric
+        :return: dictionary containing metric results on test set
+
+        """
+        if self.model.current_model is None:  # no AL steps taken so far
+            self.model.train(self.l_loader, self.valid_loader)
+
+        if test_loader is None:
+            result = self.model.test(self.test_loader)
+        else:
+            result = self.model.test(test_loader)
+        # reset current model, to avoid issues when model does not need to be trained during al_step.
+        self.model.current_model = None
+
+        if self.data_manager.top_unlabelled is not None:
+            result["hit_ratio"] = (
+                np.nan
+                if query is None
+                else len(set(query) & self.data_manager.top_unlabelled) / len(self.data_manager.top_unlabelled)
```

Comment:
maybe this should be done by a separate method to improve readability.

Reply:
Done